src/detect-file-data.c

```diff
@@ -76,12 +76,104 @@ void DetectFiledataRegister(void)
 static int DetectFiledataSetup (DetectEngineCtx *de_ctx, Signature *s, char *str)
 {
     SCEnter();
-    if ((s->init_flags & SIG_FLAG_INIT_FLOW) && (s->flags & SIG_FLAG_TOSERVER) && !(s->flags & SIG_FLAG_TOCLIENT)) {
-        SCLogError(SC_ERR_INVALID_SIGNATURE, "Can't use file_data with flow:to_server or from_client with http.");
+
+    if (!DetectProtoContainsProto(&s->proto, IPPROTO_TCP) && s->alproto != ALPROTO_UNKNOWN && s->alproto != ALPROTO_HTTP && s->alproto != ALPROTO_SMTP) {
+        SCLogError(SC_ERR_CONFLICTING_RULE_KEYWORDS, "rule contains conflicting keywords.");
         return -1;
     }
 
-    s->list = DETECT_SM_LIST_HSBDMATCH;
+    if (s->alproto == ALPROTO_HTTP || s->alproto == ALPROTO_SMTP) {
```

Comment:
Do we need the alproto check here? The behaviour is the same in both smtp and http cases, so we might just as well not check it?